### PR TITLE
Usability: Show mouse hover on all clickable Labels/Lists

### PR DESCRIPTION
**Pull request by kelvie**
Labels show the hover colour when on_click is set, HotkeyLabels should also do the same when they are clickable.

In my opinion, this makes it a lot easier to click, as someone who's not very good at clicking.

And as a usability suggestion, feel free to disagree and reject this PR!

https://user-images.githubusercontent.com/5893/218945682-c5d0dfdc-9655-4b80-951b-910be1f36630.mp4

### CODE REVIEW

**Comment by myk002** [comment]
This is a good idea, and one that is long overdue. I do wonder, though, if we should be highlighting the background instead of changing the color of the foreground. As it is, it looks like the hovered widget is being *de-emphasized* instead of highlighted.

**Comment by kelvie** [comment]
Highlighting the background would be amazing, I can play around with that, though the harder part is deciding the colour 😁

**Comment by kelvie** [comment]
One option, swap BG/FG:

https://user-images.githubusercontent.com/5893/218957888-3b960c32-52c9-45d4-b443-e3acf9b2ef38.mp4

**Comment by kelvie** [comment]
Yeah, I wish the click area could be larger, but I imagine that's impossible to do due to how it's laid out (it'd be great if we can render the buttons like in the actual UI, but those don't have hover and therefore are worse in that regard).

Or we can create a new button widget with a min height of 3 (one line above and below the text is also clickable)

**Comment by kelvie** [comment]
New patch looks like this, not sure if we can do better than this given the limited colourspace.

https://user-images.githubusercontent.com/5893/219275663-d4a97889-cdde-48fa-830d-1c4636bcb63a.mp4

**Comment by kelvie** [comment]
And notice the `blueprint` part disappears, so this isn't without regression, I think.

**Comment by kelvie** [comment]
(also `List`s also are in pretty dire need of mouseovers, as it's super easy to misclick a list item)

File: library/lua/gui/widgets.lua [hunk 1]
```diff
@@ -1594,6 +1634,23 @@ function List:init(info)
     end
 
     self.last_select_click_ms = 0 -- used to track double-clicking on an item
+
+    -- Inverts the brightness of the color
+    invert = function(color)
+        return (color + 8) % 16
+    end
```
Comment (myk002) [inline library/lua/gui/widgets.lua:1641]: this might be a useful global utility function, maybe better placed in gui.lua

**Review by myk002 — COMMENTED** [review]
looks good! though we should wait for feedback from others on discord

**Comment by kelvie** [comment]
OK, this may need a bit more testing in more dialogs (I'm not familiar with the breadth of the GUIs), but I think it's ready, if our resident artists approve

https://user-images.githubusercontent.com/5893/219561742-1d25de24-751d-49b0-82bd-a3bc5734b2d4.mp4


https://user-images.githubusercontent.com/5893/219561755-a76b3ddf-3d25-42d9-b85f-2e86dd32f0bd.mp4

**Comment by kelvie** [comment]
I also have a pretty fast PC, so I'm not sure how these affect performance, since I made some changes to some functions that are called fairly often.

File: library/lua/gui/widgets.lua [hunk 2]
```diff
@@ -1221,12 +1249,9 @@ function Label:init(args)
...
     self:addviews{self.scrollbar}
 
-    -- use existing saved text if no explicit text was specified. this avoids
-    -- overwriting pre-formatted text that subclasses may have already set
     self:setText(args.text or self.text)
-    if not self.text_hpen then
-        self.text_hpen = ((tonumber(self.text_pen) or tonumber(self.text_pen.fg) or 0) + 8) % 16
-    end
+
+    -- self.text_hpen = make_hpen(self.text_pen, self.text_hpen)
```
[the remark this answers is not in the document]
Comment (kelvie) [reply library/lua/gui/widgets.lua:1254]: Yeah I left this in to try out what @TaxiService asked about inverting per-character. I'll remove this now as it sounds like that's what we want to do (and not have a background colour).

File: library/lua/gui.lua [hunk 3]
```diff
@@ -987,4 +987,10 @@ function FramedScreen:onRenderFrame(dc, rect)
     paint_frame(dc,rect,self.frame_style,self.frame_title)
 end
 
+-- Inverts the brightness of the color, optionally taking a "bold" parameter,
+-- which you should include if you're reading the fg color of a pen.
+function invert_color(color, bold)
```
[the remark this answers is not in the document]
Comment (myk002) [reply library/lua/gui.lua:992]: ping

**Comment by myk002** [comment]
ping on remaining PR comments. I would like to get this in so more people can have eyes on the effects and complaints can arrive before we actually release with it.

**Comment by kelvie** [comment]
Made the changes, but haven't had time to properly test it right now.

**Comment by myk002** [comment]
This PR is nice and small, so we can revert if we run into trouble. I'd just like to have it active for everyone doing development right now so oddities can be noticed.

**Pull request merged**